Fixes #4802 Feat: Add Events Role Permission API for updating permissions by Admin

bhaveshAn opened the pull request:
<!--
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->
<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

 Fixes #4802 Add Events Role Permission API for updating permissions by Admin

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes <!-- use `nosetests tests/unittests` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
<!-- If an existing function does not have a docstring, please add one -->
- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
 Fixes #4802 Add Events Role Permission API for updating permissions by Admin

#### Changes proposed in this pull request:

- Added Service Schema and API
- Added Events Role Permission Schema and API
-
@poush @SaptakS @niranjan94 @maxlorenz @schedutron @Kreijstal Please review. Thanks.

https://test31038.docs.apiary.io
**Event Role Permissions : /v1/event-role-permissions** 
![screenshot from 2018-06-18 03-07-12](https://user-images.githubusercontent.com/22936570/41512360-00fa6aac-72a5-11e8-8d89-e6037a112c76.png)
![screenshot from 2018-06-18 03-07-41](https://user-images.githubusercontent.com/22936570/41512361-015a4db4-72a5-11e8-9c50-84de01dd072f.png)

**Role Relationship : /v1/event-role-permissions/1/role** 

![screenshot from 2018-06-18 03-08-34](https://user-images.githubusercontent.com/22936570/41512362-019eaec8-72a5-11e8-922e-62a1f6c21261.png)

**Service Relationship : /v1/event-role-permissions/1/service** 

![screenshot from 2018-06-18 03-09-01](https://user-images.githubusercontent.com/22936570/41512363-01e5b520-72a5-11e8-99ac-e6063c02ea55.png)

**Services List API : /v1/services** 
![screenshot from 2018-06-18 03-10-24](https://user-images.githubusercontent.com/22936570/41512381-37e89f34-72a5-11e8-8176-a0c19a215409.png)

**Services Detail API : /v1/services/1** 

![screenshot from 2018-06-18 03-10-35](https://user-images.githubusercontent.com/22936570/41512382-3849bc42-72a5-11e8-87a5-2b5d1080e689.png)

srv-twry commented on app/api/schema/event_role_permissions.py:
```diff
@@ -0,0 +1,37 @@
+from app.api.helpers.utilities import dasherize
+
+
+class EventsRolePermissionSchema(Schema):
```
When you've defined ```default``` values for some fields in the Schema, please use the ```use_defaults``` decorator. It's declared in the ```utils/common.py``` file. If you don't use it, then it won't add the default values before schema validation.

bhaveshAn replied:
Done @srv-twry Please review.

srv-twry commented on app/api/schema/services.py:
```diff
@@ -0,0 +1,25 @@
+from app.api.helpers.utilities import dasherize
+
+
+class ServiceSchema(Schema):
```
Same as above.

bhaveshAn replied:
Done @srv-twry Please review.

Kreijstal commented on app/api/service.py:
```diff
@@ -0,0 +1,26 @@
+
+class ServiceList(ResourceList):
+    """
+    List services
```
Maybe a more verbose description would be better, like what the function does or why it does it.

schedutron commented on app/api/service.py:
```diff
@@ -0,0 +1,26 @@
+                  'model': Service}
+
+
+class ServiceDetail(ResourceDetail):
```
This class is very similar to the class above. Was wondering whether it could inherit from `ServiceList`, or maybe it would be an overkill?

bhaveshAn replied:
Please see ResourceList [GET, POST] used for (getting list of objects and create new objects) and ResourceDetail [GET, PATCH, DELETE] used for (getting, updating and deleting an object based on its ID) classes in package [flask-rest-jsonAPI](http://flask-rest-jsonapi.readthedocs.io/en/latest/)

bhaveshAn replied:
@schedutron Please review.

SaptakS replied:
@schedutron they inherit different parent classes, so we can't or shouldn't inherit from `ServiceList`

bhaveshAn commented:
@Kreijstal @srv-twry @schedutron Please review.

srv-twry commented on app/api/schema/event_role_permissions.py:
```diff
@@ -0,0 +1,39 @@
+    role = Relationship(attribute='role',
+                        self_view='v1.role_detail',
+                        self_view_kwargs={'id': '<id>'},
+                        schema='RoleSchema',
```
Related view and related view kwargs ? Aren't they required ? Have you tested every endpoint if it is working or not ?

srv-twry commented on app/api/schema/event_role_permissions.py:
```diff
@@ -0,0 +1,39 @@
+                           self_view='v1.service_detail',
+                           self_view_kwargs={'id': '<id>'},
+                           related_view='v1.service_list',
+                           schema='ServiceSchema',
```
Same as above.

bhaveshAn commented:
**API `/v1/events-role-permissions`**

![screenshot from 2018-06-14 15-17-01](https://user-images.githubusercontent.com/22936570/41404932-1da68724-6fe6-11e8-9cd3-da0cdfb5f9f6.png)
![screenshot from 2018-06-14 15-16-57](https://user-images.githubusercontent.com/22936570/41404933-1de8f06e-6fe6-11e8-9f55-d440f72e28a0.png)
![screenshot from 2018-06-14 15-16-53](https://user-images.githubusercontent.com/22936570/41404935-1e3330a2-6fe6-11e8-8aad-fdbf0513db68.png)
![screenshot from 2018-06-14 15-16-49](https://user-images.githubusercontent.com/22936570/41404936-1e768686-6fe6-11e8-907b-62c171668fa0.png)


**Role and Service Relationship APIs**

![screenshot from 2018-06-14 15-18-48](https://user-images.githubusercontent.com/22936570/41404980-4228819c-6fe6-11e8-8578-ad72d253fcf5.png)
![screenshot from 2018-06-14 15-18-25](https://user-images.githubusercontent.com/22936570/41404981-4274352e-6fe6-11e8-8e8d-a16cd4598acc.png)


Will update the API blueprints in separate PR for sake of convenience of reviews. As told by @Kreijstal to make short PRs in gitter channel.
@Kreijstal @srv-twry @schedutron @iamareebjamal Please review. Thanks !

srv-twry commented on app/api/events_role_permission.py:
```diff
@@ -0,0 +1,38 @@
+    Events Role Permission detail by id
+    """
+    schema = EventsRolePermissionSchema
+    decorators = (api.has_permission('is_admin', methods="PATCH"),)
```
So the details can be seen by anyone ? You haven't add "GET" to the methods in the decorator so it won't check if the user is the admin or not in the GET request.

srv-twry replied:
@bhaveshAn Updates on this ? 
Why aren't you checking for admin access in GET request ?

srv-twry commented on app/api/events_role_permission.py:
```diff
@@ -0,0 +1,38 @@
+    decorators = (api.has_permission('is_admin', methods="GET"),)
+    schema = EventsRolePermissionSchema
+    data_layer = {'session': db.session,
+                  'model': Permission}
```
Is POST request not allowed on this endpoint ? 
If not allowed then add the methods field:
``` Python
methods = ['GET']
```

If allowed, then add it to the methods in the decorator so that we check for permissions in the POST request also.

srv-twry commented on app/api/service.py:
```diff
@@ -0,0 +1,28 @@
+    List all services i.e. microlocation, session, speaker, track, sponsor
+    """
+    decorators = (api.has_permission('is_admin', methods="GET"),)
+    methods = ['GET']
```
Same question, Is POST request not allowed on this ? If not, then it's fine but if it is then add it.

bhaveshAn replied:
No post required

srv-twry commented on app/api/events_role_permission.py:
```diff
@@ -0,0 +1,40 @@
+    """
+    schema = EventsRolePermissionSchema
+    decorators = (api.has_permission('is_admin', methods=["GET", "PATCH"]),)
+    methods = ['GET']
```
Please check it's super field. Here is the comment:
```
#: A list of methods this view can handle.
    methods = None
```

Is PATCH request allowed on this endpoint. If yes, then also add it in it. If it is not allowed then remove it from here:
`decorators = (api.has_permission('is_admin', methods=["GET", "PATCH"]),)`

srv-twry commented on app/api/events_role_permission.py:
```diff
@@ -0,0 +1,40 @@
+    Events Role Permission Relationship
+    """
+    decorators = (api.has_permission('is_admin', methods="PATCH"),)
+    methods = ['GET']
```
Same here.

bhaveshAn commented:
@srv-twry 
> Why aren't you checking for admin access in GET request ?

It doesn't required just working same as settings API .
PATCH requires admin access not GET.

bhaveshAn commented:
@niranjan94 @iamareebjamal Please review.

SaptakS commented:
@bhaveshAn I agree that `is_admin` permission isn't required for viewing the permissions.

SaptakS commented:
@bhaveshAn Please update api blueprints and tests for Services API. Rest looks good to me. Though I haven't pulled and checked.

bhaveshAn commented:
@SaptakS 
> Please update api blueprints and tests for Services API. 

I have updated API Blueprints https://test31038.docs.apiary.io and added dredd tests for Services API. Please review.

@Kreijstal 

> Can you also provide the api blueprint?

https://test31038.docs.apiary.io  Please review

Thanks!

bhaveshAn commented:
@srv-twry @Kreijstal @maxlorenz Please review

bhaveshAn commented:
@Kreijstal @maxlorenz @SaptakS Please review

bhaveshAn commented:
@SaptakS @iamareebjamal Please review and merge.Its working fine. And is a big requirement at admin section.

srv-twry commented on tests/hook_main.py:
```diff
@@ -2198,6 +2202,95 @@ def role_role_invite(transaction):
+        event_role_permission = EventRolePermissionsFactory()
+        db.session.add(event_role_permission)
+        db.session.commit()
+    """
```
Any reason for skipping this test ?

bhaveshAn replied:
Tried a lot to make them pass but not able to. At last I find to skip this for now as this feature is priority and need to be added asap at admin section. Will uncomment and fix this in another PR.

srv-twry replied:
1. I couldn't find it mentioned anywhere that you're unable to make it pass. If you have mentioned it then it's fine but otherwise please atleast **let the reviewers know when you disable tests when they are not passing**. You can argue that the reviewer should be able to see them but in a PR of 600 lines it's very difficult to notice every small details. Moreover it is the contributors responsibility to let the reviewers know about these kind of changes IMO. Also sometimes the mentors review changes just on the basis of **trust** on you. They skim over the changes to see if something is **blatantly** wrong or not and leave the minute details on you. 

2. Please add a TODO so that we can track it. 

CC: @niranjan94 @SaptakS @iamareebjamal

srv-twry commented on tests/hook_main.py:
```diff
@@ -53,6 +53,10 @@
 from app.factories.faq_type import FaqTypeFactory
 from app.factories.feedback import FeedbackFactory
+from app.factories.service import ServiceFactory
+from app.factories.event_role_permission import EventRolePermissionsFactory
```
Please remove this.

bhaveshAn commented:
@Kreijstal 
https://test31038.docs.apiary.io Please review
Thanks!

maxlorenz commented:
LGTM
